### PR TITLE
Allow Match:Regex to match metadata.

### PR DESCRIPTION
**Pull request by igorpeshansky**

### CODE REVIEW

File: src/match_regex.c [hunk 1]
```diff
@@ -295,6 +371,24 @@ static int mr_match (const data_set_t __attribute__((unused)) *ds, /* {{{ */
...
 				vl->type_instance) == FC_MATCH_NO_MATCH)
 		return (nomatch_value);
+	if (vl->meta != NULL)
+	{
+		for (llentry_t *e = llist_head(m->meta); e != NULL; e = e->next)
+		{
+			mr_regex_t *meta_re = (mr_regex_t *) e->value;
+			char *value;
+			int status = meta_data_get_string (vl->meta, e->key, &value);
+			if (status == 0)  /* key is present */
```
Comment (octo) [inline src/match_regex.c]: Please do:

```c
if (status != 0) /* key is not present */
  continue;
/* remainder of code */
```

File: src/match_regex.c [hunk 2]
```diff
@@ -93,6 +96,12 @@ static void mr_free_match (mr_match_t *m) /* {{{ */
 	mr_free_regex (m->plugin_instance);
 	mr_free_regex (m->type);
 	mr_free_regex (m->type_instance);
+	for (llentry_t *e = llist_head(m->meta); e != NULL; e = e->next)
+	{
+		free (e->key);
```
Comment (octo) [inline src/match_regex.c]: Please search and replace all `free()` with `sfree()`. Ruben just did a tree-wide unification and I'd prefer not to muddy the waters again.

File: src/match_regex.c [hunk 3]
```diff
@@ -184,8 +187,78 @@ static int mr_config_add_regex (mr_regex_t **re_head, /* {{{ */
...
+		{
+			log_err ("mr_config_add_meta_regex: llentry_create failed.");
+			free (key);
+			return (-1);
+		}
+		/* key and entry will now be freed by mr_free_match(). */
+		llist_append (*meta, entry);
+	}
+
+	snprintf (buffer, sizeof (buffer), "%s `%s'", ci->key, key);
```
Comment (octo) [inline src/match_regex.c]: At your option: I find `ci->key` and `key` a bit hard to read. How about renaming `key` to `meta_name`?

File: src/match_regex.c (hunk 1 above)
Comment (octo) [inline src/match_regex.c]: Is this actually the correct behavior? If I set `MetaData "foo" "^bar"` and a metric does *not* have the "foo" meta data at all, I'd expect the that to be a no-match.

**Comment by octo** [comment]
Thanks @igorpeshansky, great work as usual :)

File: src/match_regex.c (hunk 1 above)
Comment (igorpeshansky) [inline src/match_regex.c]: Good catch, this felt wrong to me as well. Fixed.

File: src/match_regex.c (hunk 3 above)
Comment (igorpeshansky) [inline src/match_regex.c]: I renamed to meta_key instead, to be consistent with meta_data naming.

**Comment by igorpeshansky** [comment]
Thank you for the review, @octo! I've addressed your comments and pushed a new commit. Would you like me to squash them at some point?

**Comment by octo** [comment]
Thanks @igorpeshansky! I spotted a use of `snprintf()` I missed on the previous iteration and fixed that to avoid having another round-trip. I'll merge as soon as the build status is available and successful.

**Pull request merged**